File: patterns/creational/singleton.py
"""
Singleton Pattern - Python

Ensures a class has only one instance and provides a global point of access to it.

Use cases:
- Database connections
- Logger instances
- Configuration managers
- Cache managers
"""


class Singleton:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Singleton, cls).__new__(cls)
            cls._instance.data = "Singleton instance data"
        return cls._instance
    
    def get_data(self):
        return self.data
    
    def set_data(self, data):
        self.data = data


# Alternative: Using a decorator
def singleton(cls):
    instances = {}
    
    def get_instance(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]
    
    return get_instance


@singleton
class DatabaseConnection:
    def __init__(self):
        self.connection_string = "database://localhost"
    
    def connect(self):
        return f"Connected to {self.connection_string}"


# Example usage
if __name__ == "__main__":
    print("=== Singleton Pattern ===\n")
    
    singleton1 = Singleton()
    singleton2 = Singleton()
    
    print("Are both instances the same?", singleton1 is singleton2)  # True
    print("Instance 1 data:", singleton1.get_data())
    singleton2.set_data("Modified data")
    print("Instance 1 data after modifying instance 2:", singleton1.get_data())
    
    # Decorator-based singleton
    db1 = DatabaseConnection()
    db2 = DatabaseConnection()
    print("Database instances are the same:", db1 is db2)  # True
    print(db1.connect())

File: patterns/creational/singleton.js
/**
 * Singleton Pattern - JavaScript
 * 
 * Ensures a class has only one instance and provides a global point of access to it.
 * 
 * Use cases:
 * - Database connections
 * - Logger instances
 * - Configuration managers
 * - Cache managers
 */

class Singleton {
    constructor() {
        if (Singleton.instance) {
            return Singleton.instance;
        }
        
        this.data = 'Singleton instance data';
        Singleton.instance = this;
    }
    
    getData() {
        return this.data;
    }
    
    setData(data) {
        this.data = data;
    }
}

// Alternative: Using a function
function createSingleton() {
    let instance = null;
    
    return function() {
        if (!instance) {
            instance = {
                data: 'Singleton instance data',
                getData() {
                    return this.data;
                },
                setData(data) {
                    this.data = data;
                }
            };
        }
        return instance;
    };
}

// Example usage
console.log('=== Singleton Pattern ===\n');

const singleton1 = new Singleton();
const singleton2 = new Singleton();

console.log('Are both instances the same?', singleton1 === singleton2); // true
console.log('Instance 1 data:', singleton1.getData());
singleton2.setData('Modified data');
console.log('Instance 1 data after modifying instance 2:', singleton1.getData());

// Function-based singleton
const getSingleton = createSingleton();
const instance1 = getSingleton();
const instance2 = getSingleton();
console.log('Function-based singleton:', instance1 === instance2); // true

module.exports = { Singleton, createSingleton };

